test/Confluent.Kafka.IntegrationTests/Tests/Timestamps.cs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2016-2017 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+#pragma warning disable xUnit1026
+
+using System;
+using System.Text;
+using System.Collections.Generic;
+using Xunit;
+using Confluent.Kafka.Serialization;
+
+
+namespace Confluent.Kafka.IntegrationTests
+{
+    public static partial class Tests
+    {
+        class DeliveryHandler_TCB : IDeliveryHandler<Null, string>
+        {
+            public static List<Message<Null, string>> drs 
+                = new List<Message<Null, string>>();
+
+            public bool MarshalData
+                => true;
+
+            public void HandleDeliveryReport(Message<Null, string> message)
+            {
+                drs.Add(message);
+            }
+        }
+
+        class DeliveryHandler_TCB_2 : IDeliveryHandler
+        {
+            public static List<Message> drs 
+                = new List<Message>();
+
+            public bool MarshalData
+                => true;
+
+            public void HandleDeliveryReport(Message message)
+            {
+                drs.Add(message);
+            }
+        }
+
+        /// <summary>
+        ///     Integration tests for Producing / consuming timestamps.
+        /// </summary>
+        [Theory, MemberData(nameof(KafkaParameters))]
+        public static void CustomTimestampTests(string bootstrapServers, string singlePartitionTopic, string partitionedTopic)
+        {
+            var producerConfig = new Dictionary<string, object>
+            {
+                { "bootstrap.servers", bootstrapServers }
+            };
+
+            var consumerConfig = new Dictionary<string, object>
+            {
+                { "group.id", Guid.NewGuid().ToString() },
+                { "bootstrap.servers", bootstrapServers },
+                { "session.timeout.ms", 6000 }
+            };
+
+            List<Message<Null, string>> drs = new List<Message<Null, string>>();
+            using (var producer = new Producer<Null, string>(producerConfig, null, new StringSerializer(Encoding.UTF8)))
+            {
+                drs.Add(producer.ProduceAsync(singlePartitionTopic, null, "testvalue").Result);
+                
+                // Note: in the Message<K,V> case, timestamps of type LogAppendTime and NotAvailable are not considered errors 
```

Comment:
does this seem reasonable?

Reply:
Not sure I follow the reasoning here: how would retries affect timestamps?

Reply:
I mean manually re-trying. I can imagine a scenario where you have some hybrid allow-librdkafka-to-retry / manually get involved (+ possibly do something other than retry). I guess that is advanced usage, and I don't exactly know what good practice is / what this would look like. Perhaps there is no good argument for ever doing this, in which case the behavior as implement is probably not what we want.

If people are not likely to want to produce a `Message` that was returned as a delivery report, we should consider not having these overloads. I guess another usage though is if people want to prepare a message in one function, then produce it in another. In that case, the message overload is of course very handy. This pattern is used in the JMS Client for example, but I don't have a good feel for how prevalent it would be in general. Anyway, under this usage, we should throw an exception if timestamps of type LogAppendTime or NotAvailable are used.